### DISCUSSION TITLE
Receiving input from shell?

### DISCUSSION THREAD

**Discussion by maxbrito500** [dani007200964/Shellminator#13] [General]
Hello,

Thank you for making Shellminator available, very good piece of work. I'll soon showcase how it is being used on my project and will try to keep contributing with documentation and small fixes/improvements where possible.

In the meanwhile have a question: How can we accept user input for commands?

For example, we need a command to format the TF card and need a positive confirmation of the action since it has destructive results. Therefore the end user needs to press "Y" to go ahead and format the disk. How would this be possible to achieve with shellminator?

Thanks!

**Comment by dani007200964** [comment]
There are two use cases for this problem:

1. If you use the execution function instead of Commander, you have to know which peripheral you use as a channel and manually poll information from it. For example, if you use Serial as a channel and you want to confirm something with a 'Y' character:

```c
void executionFunction( char* command ){
    while( !Serial.available() ){
         // Waiting for user input
    }

    char c = Serial.read();

    if( c == 'y' || c == 'Y' ){
        // Allowed to do something
    }

    else{
        // It is forbidden to do something
    }
}
```

2. If you use Commander it is a bit more advanced, but it can automate a lot of things. In this case, every command will be called with the reference to its source channel. This way you don't have to track the origin of the message.

```c
void delete_func(char *args, Stream *response ){
    while( !response -> available() ){
         // Waiting for user input
    }

    char c = response -> read();

    if( c == 'y' || c == 'Y' ){
        // Allowed to do something
    }

    else{
        // It is forbidden to do something
    }
}
```

I thought about making some functions within Shellminator to read a line or a prompt to make this task easy, but it is a bit of time hungry task. If you need this functionality it is possible to implement it in the future, but please leave a feature request. There is a template for it in the issue section. I think I can make it with a few tweaks depending on my free time.

**Comment by dani007200964** [comment]
> I'll soon showcase how it is being used on my project and will try to keep contributing with documentation and small fixes/improvements where possible.

Good luck with your project! If you need some help, please let me know 🙂
Also thank you for your work! I really appreciate it!

**Comment by maxbrito500** [comment]
Thank you so much for the quick reply!

Your code reply is already a good start. In this case the connection is through WiFi, so it would be better to keep the connection type agnostic. I've tried the code inside the "reboot" command, however it just passes right through without waiting for user input. I've made a video showing the behaviour:
[![asciicast](https://asciinema.org/a/UhPmjDQe3Fx5oVntfWBn885g5.svg)](https://asciinema.org/a/UhPmjDQe3Fx5oVntfWBn885g5)

The code itself:
```
// reboot the device
void func_reboot(char *args, Stream *response ){

    response -> println("Reboot now? (Y/n)");
    while( !response -> available() ){
         // Waiting for user input
    }

    char c = response -> read();

    if( c == 'y' || c == 'Y' ){
        response -> println("System is rebooting NOW");
        esp_restart();
    }
}
```

Would say that there is no need yet to include the read function in easier form inside shellminator, the current code is already simple enough to use. I just don't know why it isn't waiting for input. Do you have any ideas?

Thank you!

btw. I'll keep contributing with documentation and the other linux-like commands that are being written in case there is interest.

**Reply by dani007200964** [reply]
Interesting. I have a few ideas. The TCP packets have to be processed, but I'm not sure why it passes through the available function. If I finish work I will check it. Can you tell me what is the exact setup you used( ESP variant IDE ... etc. )?

It has to work with Serial, I checked it, but TCP and network stuff is a bit more advanced. I worked on the Telnet implementation for a month to make it work. I think I could use some implementations from that part of the code, but testing this stuff is a nightmare.

> btw. I'll keep contributing with documentation and the other linux-like commands that are being written in case there is interest.

That would be really good! Thank you :)

**Comment by maxbrito500** [comment]
> The TCP packets have to be processed

Could be. Quite tricky to pinpoint.

> Can you tell me what is the exact setup you used( ESP variant IDE ... etc. )?

I'm using the Sunton board ESP32_2432S028R that is available at https://www.aliexpress.com/item/1005004502250619.html and this is a ESP-WROOM-32 machine. The project is developed with PlatformIO. You find the project source code here: https://github.com/radio3-network/BluemchenGUI/blob/main/src/core/blum_terminal.cpp

> I worked on the Telnet implementation for a month to make it work. I think I could use some implementations from that part of the code, but testing this stuff is a nightmare.

Can certainly imagine. I'm a long time Java developer starting now with Arduino with C++, what a difference! Now I understand that java development has it really easy by comparison to developing embedded things. 😅

**Reply by dani007200964** [reply]
I think I have some good news. The problem is caused by some leftover characters in the input buffer. The solution was to empty the input buffer before calling the execution function or Commander.  Looks like the TCP stack is running on a different thread, so far so good. I also added two static functions, so they can be called from the outside:
* __waitForKey__ is good for things you expected. It waits for single or multiple keys and returns if it was successful.
* __input__ is a nice way to read until the return key is pressed. It returns how many characters have been collected.

Both of them can be protected with a timeout. I created an example for these functions it can be found [here](https://github.com/dani007200964/Shellminator/blob/dev/examples/Shellminator_ESP32_server_with_Commander_prompt/Shellminator_ESP32_server_with_Commander_prompt.ino)

However it is on the __dev__ branch, so you have to install it manually to PlatformIO. Also in this branch, I redesigned the whole state machine to make it more flexible for later upgrades. Currently, I did not have much time to test the new version, so It will be probably buggy.

If you have some spare time you can test it. If you find anything interesting, please let me know 😃

**Comment by maxbrito500** [comment]
Hey, this is fantastic! THANK YOU.

I'll happily test and report back the results. Very good decision with waitForKey and input, both them are really nice for interactive shells. On the free time of my free time I'm writing a text game engine: https://github.com/radio3-network/storycraft/blob/main/tales/forest/main.tale I will later test that game through shellminator.

Again, thank you so much.

**Reply by dani007200964** [reply]
Good luck with the game! When you finish it, please let me know! I always wanted to create some kind of table, that shows the projects, that use Sellminator.

**Comment by maxbrito500** [comment]
Hello, I've downloaded the dev branch and replaced the library.

Can confirm the same results as the demonstration source code. These are working: key selection and timeout.

Did not yet experimented yet reading the chars or stress testing for unexpected keystrokes but should try in next weeks as the code gets more developed and can test the input more intensively.

**Reply by dani007200964** [reply]
Sounds great! Good luck!
